contracts/WildcardsQV.sol

```diff
@@ -41,83 +61,119 @@ contract WildcardsQV is Initializable {
+    function createProposal(address payable _addressOfCharity)
         external
         onlyAdmin
         returns (uint256 newProposalId)
     {
-        // So the first proposal will have an ID of 1
-        proposalId = proposalId.add(1);
-
-        proposalDetails[proposalId] = proposalHash;
+        proposalAddresses[proposalId] = _addressOfCharity;
         //proposalOwner[proposalId] = msg.sender;
         //benefactorsProposal[msg.sender] = proposalId;
         state[proposalId] = ProposalState.Active;
+        emit LogProposalCreated(proposalId,_addressOfCharity);
+        // So the first proposal will have an ID of 0
+        proposalId = proposalId.add(1);
         return proposalId;
     }
 
     ///////////////////////////////////
     /////// vote function  ///////////
     ///////////////////////////////////
-    function vote(uint256 proposalIdToVoteFor, uint256 amount) external {
+    function vote(uint256 proposalIdToVoteFor, uint256 amount, uint256 sqrt) external {
         // Check they are a wildcards user
-        // Check they have wildcards loyalty Tokens
-        // Check they are voting for a valid proposal + the they haven't yet voted for the proposal in this iteration
-
-        // Approve us to send wildcards tokens on their behalf
+        require(wildCardToken.balanceOf(msg.sender)>0, "Does not own a WildCard");
+        // Check they have at least 1 wildcards loyalty token:
+        require(amount >= 10**18, " Minimum vote one token");
+        // Check they are voting for a valid proposal: 
+        require(state[proposalIdToVoteFor] == ProposalState.Active, "Proposal not Active");
+        // Check that they haven't yet voted for the proposal in this iteration
+        require(!hasUserVotedForProposalIteration[proposalIteration][msg.sender][proposalIdToVoteFor], "Already voted on this proposal");
         // Send their wildcards tokens to the burn address
-
+        require(loyaltyToken.transferFrom(msg.sender,burnAddress,amount), "Loyalty Token transfer failed");
+        // Validate the square root
+        require(sqrt.mul(sqrt) == amount, "Square root incorrect");
         // Add the tokens to the total tally
         proposalVotes[proposalIteration][proposalIdToVoteFor] = proposalVotes[proposalIteration][proposalIdToVoteFor]
-            .add(sqrt(amount));
+            .add(sqrt);
+        uint256 _currentVoteCount = proposalVotes[proposalIteration][proposalIdToVoteFor];
+        // Take note that the user has voted on this proposal so they can't do it again
+        hasUserVotedForProposalIteration[proposalIteration][msg.sender][proposalIdToVoteFor] = true;
+        // Update currentWinner, currentHighestVoteCount, totalVotes
+        if (_currentVoteCount > currentHighestVoteCount) {
+            currentHighestVoteCount = _currentVoteCount;
+            currentWinner = proposalIdToVoteFor;
+        }
+        totalVotes = totalVotes.add(_currentVoteCount);
```

Comment:
This will double count some votes :) Since _currentVoteCount takes into account all the votes for that proposal for that iteration, instead of just the votes specific to this iteration. Going to make a PR with one or two small changes to fix this